Getting rid of the conn suite

SimoneDutto opened the pull request:
# Description

Approach to get rid of the conn suite.

Rationale: keep most of the test suites, and setup as we currently have but gently switch from the connsuite to an externally bootstrapped controller.

Outline:
- the juju controller is bootstrapped via the Juju cli and configured to use the same jwks is used by the test's jimm
- jimm's tests will still create a running jimm per test via the test harness as usual and they'll attach the externally bootstrapped controller.
- jimm will run command on the backing juju controller

# JWKS
I've added an alpine nginx to serve a fixed jwks.json file that the backing juju controller will be bootstrapped with.
`JWKS_DNS=jwks.localhost CONTROLLER_NAME=test-e2e ./local/jimm/setup-controller.sh`

And reuse the same jwks setting up JIMM inside the `JIMMSuite`

# Adding the controller
Right now in the E2ESuite we parse the yaml config file defined at JIMM_BACKING_CONTROLLER_CONFIG path.
And we add the controller.
For now it's just a single controller but the code is ready to add multiple controllers via the same controllers.yaml file.

```
s.AddController(c, "controller-1", &api.Info{
		ControllerUUID: "b2746812-29be-47a6-8cae-40717c390a19",
		Addrs:          []string{"10.165.178.29:17070"},
		Tag:            names.NewUserTag("admin"),
		Password:       "",
		CACert:         ""
	})
```
This file can be generated via `make generate-test-env <controller>`

# Credentials
Here the approach is similar to the tf provider, we list the credentials on the client and we add to JIMM from the filestore.

# Change to the tests
- using the E2ESuite instead of the BootstrapSuite
- remove all ConnSuite specific calls. Ex.
` f := factory.NewFactory(stateMachine3.State, s.StatePool) `
and calling the JIMM's client directly, that will target the backing controller.
- do proper cleanup. Ex. adding a model to the backing controller, now requires cleanup because there isn't a controller per test.

# Downsides

- Bootstrapping multiple controllers will require a new suite called "E2EMultipleControllersSuite" and a different way to pass the connection details (maybe a yaml file).
- Bootstrapping outside the test harness, so there is no control over isolation for tests requiring a full controller.

kian99 commented on docker-compose.yaml:
```diff
@@ -152,6 +152,18 @@ services:
       traefik.http.routers.juju-dashboard.tls: true
       traefik.http.services.juju-dashboard.loadbalancer.server.port: 8080
 
+  jwks-server:
```
A comment above this line about what the purpose of this container would be helpful, or maybe better would be an update to `local/README.md`

kian99 commented on testing/modelmanager_test.go:
```diff
@@ -0,0 +1,68 @@
+
+var _ = gc.Suite(&modelE2eManagerSuite{})
+
+func (s *modelE2eManagerSuite) TestListModelSummaries(c *gc.C) {
```
What do you think about moving this test outside of `internal/jujuapi`. The spec states we'll write these new tests in `testing/e2e/`, a nice step towards integration tests that don't need to reach for internals.

SimoneDutto committed:
feat(remove-conn-suite): remove the connsuite in favor of bootstrapping a controller externally

In this pr we start the removal of the ConnSuite by modifying current suites to support an
externally bootstrapped controller.

SimoneDutto commented on .vscode/settings.json:
the reason why we can't track .vscode anymore it's described in `testing/README.md` at the bottom.
Basically, babak's extension doesn't support ${workspaceFolder}` expansion and `go.testEnvFile`, so we can't really track this file no more. Because everybody here will have a different absolute path for when he wants to run the E2E_test

luci1900 replied:
It's fairly common to commit a `settings.example.json` or similar. That way everyone can have their own, but it's easier for new joiners to set up.

SimoneDutto replied:
yeah, added

luci1900 reviewed (approved):
I like this approach.

kian99 commented on testing/modelmanager_test.go:
Name of the file needs fixing

SimoneDutto replied:
fixed

kian99 commented on testing/package_test.go:
```diff
@@ -0,0 +1,21 @@
+
+func onlyRunE2ETests(c *gc.C) {
+	if _, ok := os.LookupEnv("RUN_E2E_TESTS"); !ok {
+		c.Skip("Skipping e2e tests. Set RUN_E2E_TESTS=true to run them.")
```
Can we move this check into `TestPackage`?

SimoneDutto replied:
i've tried it but it wasn't working. I'll retry

SimoneDutto replied:
it works

kian99 commented on .vscode/.settings.json.example:
Should we add the .settings file to .gitignore now since everyone will need a different config potentially?

SimoneDutto replied:
this is just the .example, it's not used by vscode

kian99 replied:
What I mean is, the other file should not be commited anymore since it will need to contain user specific paths. We could add it to gitignore, if it's not already there.

SimoneDutto replied:
in fact i've deleted the `settings.json` and added to .gitignore

kian99 replied:
Ah I see that now thanks!

kian99 commented on testing/README.md:
````diff
@@ -0,0 +1,48 @@
+   Start the dev environment to run these tests:
+
+   ```bash
+   make test-env
````
Shouldn't this be dev-env?

SimoneDutto replied:
i think this is the right one, because we don't need jimm

kian99 commented on testing/README.md:
Seems like we could create a make target like `e2e_test` that does these 3 steps as a convenience.

SimoneDutto replied:
it's just that you don't want to do all 3 at all time, and i prefer them to be separated for now. We can create it later

kian99 reviewed (commented):
Thanks for this
